src/Voron/Impl/LowLevelTransaction.cs

```diff
@@ -1132,14 +1133,19 @@ public void EnsurePagerStateReference(PagerState state)
             if (state == _lastState || state == null)
                 return;
 
-            if (_pagerStates.Add(state) == false)
+            _forTestingPurposes?.ActionToCallDuringEnsurePagerStateReference?.Invoke();
+
+            if(_pagerStates.Contains(state))
             {
                 _lastState = state;
                 return;
             }
 
             state = state.CurrentPager.GetPagerStateAndAddRefAtomically(); // state might hold released pagerState, and we want to add ref to the current (i.e. data file was re-allocated and a new state is now available). RavenDB-6950
+
             _lastState = state;
+
+            _pagerStates.Add(state);
```

Comment:
Fix is here - we didn't add `state` returned by `GetPagerStateAndAddRefAtomically()` which takes into account that we have a new state